handler/src/main/java/io/netty/handler/ssl/DefaultOpenSslSession.java

```diff
@@ -0,0 +1,349 @@
+        this.localCertificateChain = localCertificate;
+    }
+
+    @Override
+    public void setPacketBufferSize(int packetBufferSize) {
+        this.packetBufferSize = packetBufferSize;
+    }
+
+    @Override
+    public void updateLastAccessedTime() {
+        lastAccessed = System.currentTimeMillis();
+    }
+
+    @Override
+    public Certificate[] getPeerCertificates() throws SSLPeerUnverifiedException {
+        if (SslUtils.isEmpty(peerCerts)) {
+            throw new SSLPeerUnverifiedException("peer not verified");
+        }
+        return peerCerts.clone();
+    }
+
+    @Override
+    public Certificate[] getLocalCertificates() {
+        Certificate[] localCerts = localCertificateChain;
+        if (localCerts == null) {
+            return null;
+        }
+        return localCerts.clone();
+    }
+
+    @Override
+    public X509Certificate[] getPeerCertificateChain() throws SSLPeerUnverifiedException {
+        if (SslUtils.isEmpty(x509PeerCerts)) {
+            throw new SSLPeerUnverifiedException("peer not verified");
+        }
+        return x509PeerCerts.clone();
+    }
+
+    @Override
+    public Principal getPeerPrincipal() throws SSLPeerUnverifiedException {
+        Certificate[] peer = getPeerCertificates();
+        // No need for null or length > 0 is needed as this is done in getPeerCertificates()
+        // already.
+        return ((java.security.cert.X509Certificate) peer[0]).getSubjectX500Principal();
+    }
+
+    @Override
+    public Principal getLocalPrincipal() {
+        Certificate[] local = localCertificateChain;
+        if (SslUtils.isEmpty(local)) {
+            return null;
+        }
+        return ((java.security.cert.X509Certificate) local[0]).getIssuerX500Principal();
+    }
+
+    @Override
+    public String getCipherSuite() {
+        return cipher;
+    }
+
+    @Override
+    public String getProtocol() {
+        return protocol;
+    }
+
+    @Override
+    public int getPacketBufferSize() {
+        return packetBufferSize;
+    }
+
+    @Override
+    public int getApplicationBufferSize() {
+        return applicationBufferSize;
+    }
+
+    @Override
+    public void tryExpandApplicationBufferSize(int packetLengthDataOnly) {
+        if (packetLengthDataOnly > ReferenceCountedOpenSslEngine.MAX_PLAINTEXT_LENGTH &&
+                applicationBufferSize != ReferenceCountedOpenSslEngine.MAX_RECORD_SIZE) {
+            applicationBufferSize = ReferenceCountedOpenSslEngine.MAX_RECORD_SIZE;
+        }
+    }
+
+    @Override
+    protected void deallocate() {
+        if (sslSession != -1) {
+            SSLSession.free(sslSession);
+        }
+    }
+
+    @Override
+    public DefaultOpenSslSession touch() {
+        super.touch();
+        return this;
+    }
+
+    @Override
+    public DefaultOpenSslSession touch(Object hint) {
+        return this;
+    }
+
+    @Override
+    public DefaultOpenSslSession retain() {
+        super.retain();
```

Comment:
If this a ref-counted object, should `retain()` also perform `io.netty.internal.tcnative.SSLSession.upRef`?

Reply:
nope... We only reference count the java object itself to simplify things here.